File: Python/123__Best_Time_to_Buy_and_Sell_Stock_III.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) <= 1:
            return 0

        # O(n) counting from left, find the max gain up to each day (not ending at each day)
        left = [0] * len(prices)
        curmin = prices[0]
        for i in range(1, len(prices)):
            curmin = min(curmin, prices[i])
            left[i] = max(prices[i] - curmin, left[i - 1])

        # O(n) counting from right
        right = [0] * len(prices)
        curmax = prices[-1]
        for i in range(len(prices) - 2, -1, -1):
            curmax = max(curmax, prices[i])
            right[i] = max(curmax - prices[i], right[i + 1])

        # O(n)
        result = 0
        for i in range(len(prices)):
            result = max(result, left[i] + right[i])
        return result

File: Java/123__Best_Time_to_Buy_and_Sell_Stock_III.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) return 0;
        int result = -1;
        int len = prices.length;
        int[] left = new int[len];
        int[] right = new int[len];
        int cur_min = prices[0], cur_max = prices[len - 1];
        for (int i = 1; i < len; i++) {
            cur_min = Math.min(cur_min, prices[i]);
            left[i] = Math.max(left[i - 1], prices[i] - cur_min);
        }
        for (int i = len - 2; i >= 0; i--) {
            cur_max = Math.max(cur_max, prices[i]);
            right[i] = Math.max(right[i + 1], cur_max - prices[i]);
        }
        for (int i = 0; i < len; i++)
            result = Math.max(result, left[i] + right[i]);
        return result;
    }
}
